---
title: "SeaDNA 12S rRNA reference library coverage"
author: "Rupert A. Collins"
date: "`r format(Sys.time(), '%d %B %Y')`"
output: github_document
---


##### Methods and description
This document describes the current sampling for the 12S rRNA fish reference library for the SeaDNA project. 
The document is a dynamic knitr document and can be updated quickly using the Makefile in SeaDNA/scripts.
A list of species from the UK was generated from three sources: GBIF, FishBase, and the Water Framework Directive list of transitional species.
This list was filtered to identify synonyms and duplicates, and annotated with FishBase taxonomic classification and FishBase common names.
Next a sub-list of "common" species was generated. 
These were species that we believe are likely to be encountered in eDNA surveys of inshore and transitional waters of the UK, and comprise most of the species in Henderson (2015).
Most of the remaining are either introduced species, rarely encountered migrants, oceanic pelagics, or deep sea organisms.


To calculate coverage we used the Bristol SeaDNA tissue catalogue, and also performed a search of the GenBank/BOLD databases. 
Because of inconsistencies in how researchers annotate their GenBank submissions and the differing internal coverage of primer pairs for particular gene fragments, we performed a search requesting all mitochondrial DNA.
Then we pulled out the ~170 bp Miya fragment from all the mtDNA using a hidden markov model. This enabled us to have greater confidence that useful sequences had not been missed. 
For the resulting sequences we then tabulated all their metadata from GenBank in order to allow us to curate a custom reference library according to various criteria (e.g. must have reference specimen or locality data).


```{r load_libs, include=FALSE}
# load up libs
require("tidyverse")
require("magrittr")
require("knitr")
require("rfishbase")
```


```{r load_data, include=FALSE}
# load up the data
tissues.df <- read_csv(file="../species/tissues-master.csv")
source("https://raw.githubusercontent.com/boopsboops/reference-libraries/master/scripts/references-load.R")
genbank.df <- reflib.orig
activity.df <- read_csv(file="../references/activity-dates.csv")
```


```{r activities, include=FALSE}
# get dates and genbank version
gb.dl <- activity.df %>% filter(activity=="download mtDNA all uk species")
gb.version <- gb.dl$version
gb.date <- gb.dl$date
```


```{r tissue_clean, include=FALSE}
# clean up the tissues data and validate names
tissues.filt <- tissues.df %>%
    mutate(sciName=paste(genus, specificEpithet)) %>%
    filter(class!="Malacostraca") %>%
    filter(!is.na(specificEpithet)) %>% 
    mutate(sciNameValid=pull(uk.species.table.orig,validName)[match(sciName,pull(uk.species.table.orig,speciesName))])
#
tissues.sp <- tissues.filt %>%
    select(sciNameValid) %>%
    count(sciNameValid) %>%
    rename(tissueCount=n)
```


```{r genbank_clean, include=FALSE}
# clean up the GenBank data
genbank.df %<>% filter(!is.na(nucleotidesFrag.12s.miya.noprimers))
gb.dim <- genbank.df %>% summarise(n=n())

genbank.sp <- genbank.df %>%
    select(sciNameValid) %>%
    count(sciNameValid) %>%
    rename(genbankCount=n)
```


```{r species_clean, include=FALSE}
# clean up the species data
uk.species.table.val <- uk.species.table %>%
    rename(sciNameValid=validName)
```


```{r tables_join, include=FALSE}
# join to the uk species table
combined.df <- dplyr::left_join(uk.species.table.val, tissues.sp, by="sciNameValid")
combined.df <- dplyr::left_join(combined.df, genbank.sp, by="sciNameValid")
combined.df %<>% mutate(tissueCount=as.integer(replace_na(tissueCount,0)), genbankCount=as.integer(replace_na(genbankCount,0)))
```


```{r common_subset, include=FALSE}
# subset the common species
common.df <- combined.df %>%
    filter(commonSpecies==TRUE) %>%
    arrange(class, order, family, sciNameValid) %>%
    select(family, sciNameValid, commonName, tissueCount, genbankCount)
```


```{r rare_subset, include=FALSE}
# subset the rare species
rare.df <- combined.df %>%
    filter(commonSpecies==FALSE) %>%
    arrange(class, order, family, sciNameValid) %>%
    select(family, sciNameValid, commonName, tissueCount, genbankCount)
```


```{r needed_subset, include=FALSE}
# subset the common species with no data
needed.species <- common.df %>%
    filter(tissueCount==0 & genbankCount==0) %>%
    select(family, sciNameValid, commonName)
```


```{r percentages, include=FALSE}
# calculate percentages for each group
uk.all <- combined.df %>% count()
got.com <- combined.df %>% filter(commonSpecies==TRUE) %>% filter(tissueCount>0 | genbankCount>0) %>% count()
tot.com <- combined.df %>% filter(commonSpecies==TRUE) %>% count()
got.com.prop <- paste0(round(((got.com$n/tot.com$n)*100), digits=0), "%")
#
got.all <- combined.df %>% filter(commonSpecies==FALSE) %>% filter(tissueCount>0 | genbankCount>0) %>% count()
tot.all <- combined.df %>% filter(commonSpecies==FALSE) %>% count()
got.all.prop <- paste0(round(((got.all$n/tot.all$n)*100), digits=0), "%")
#
med.com <- combined.df %>% filter(commonSpecies==TRUE) %>% mutate(tots=tissueCount+genbankCount) %>% summarise(m=median(tots))
singles <- combined.df %>% filter(commonSpecies==TRUE) %>% mutate(tots=tissueCount+genbankCount) %>% filter(tots==1) %>% count()
singles.prop <- paste0(round(((singles$n/tot.com$n)*100), digits=0), "%")
#
tis.n <- tissues.filt %>% select(sciNameValid) %>% count()
tis.sp <- tissues.filt %>% select(sciNameValid) %>% distinct() %>% count()
tis.sp.loc <- tissues.filt %>% filter(!is.na(locality)) %>% select(sciNameValid) %>% distinct() %>% count()
```

##### Results
The total number of UK species is estimated to be around `r uk.all`.
GenBank and BOLD were searched on `r gb.date` (GenBank version `r gb.version`).
A total of `r gb.dim` records were recovered.
In terms of tissue sampling, `r tis.n` samples have been taken from `r tis.sp` fish species identified to species level. 
Of these, `r tis.sp.loc` species have locality data. 
Combined with the GenBank sequence data, `r got.com.prop` of the `r tot.com` common species have been sampled (Table 2), and `r got.all.prop` of the `r tot.all` rare species (Table 3).
For the common species, the median number of tissues or GenBank sequences per species is `r med.com`, while `r singles` (`r singles.prop`) are represented by only one tissue or sequence.
Table 1 lists the species of the highest priority to source.


###### Table 1. Highest priority species (common species with no tissue sample or GenBank data).
```{r print_needed, echo=FALSE, results="asis"}
# print the needed species table
needed.species %>%
    mutate(sciNameValid=str_replace_all(sciNameValid, pattern="$|^", replacement="*")) %>%
    rename(Family=family, `Scientific Name`=sciNameValid, `Common Name`=commonName) %>%
kable()
```
---


###### Table 2. All common UK species with counts for tissue samples collected and sequence data obtained from GenBank (number of individuals).
```{r print_common, echo=FALSE, results="asis"}
options(knitr.kable.NA="")
# print the common species table
common.df %>%
    mutate(sciNameValid=str_replace_all(sciNameValid, pattern="$|^", replacement="*")) %>%
    mutate(tissueCount=replace(tissueCount,which(tissueCount==0),NA), genbankCount=replace(genbankCount,which(genbankCount==0),NA)) %>%
    rename(Family=family, `Scientific Name`=sciNameValid, `Common Name`=commonName, `Tissue Count`=tissueCount, `GenBank Count`=genbankCount) %>%
kable()
```
---


###### Table 3. All other UK species (reported in UK waters, but not listed in common species).
```{r print_rare, echo=FALSE, results="asis"}
options(knitr.kable.NA="")
# print the all UK species table
rare.df %>%
    mutate(sciNameValid=str_replace_all(sciNameValid, pattern="$|^", replacement="*")) %>%
    mutate(tissueCount=replace(tissueCount,which(tissueCount==0),NA), genbankCount=replace(genbankCount,which(genbankCount==0),NA)) %>%
    rename(Family=family, `Scientific Name`=sciNameValid, `Common Name`=commonName, `Tissue Count`=tissueCount, `GenBank Count`=genbankCount) %>%
kable()
```

